File: app/src/main/java/com/mx/gillustrated/activity/MainActivity.java
package com.mx.gillustrated.activity;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import com.mx.gillustrated.common.DBCall;
import com.mx.gillustrated.common.MConfig;
import com.mx.gillustrated.component.MainActivityHeader;
import com.mx.gillustrated.component.MainActivityListView;
import com.mx.gillustrated.component.MainActivityTop;
import com.mx.gillustrated.dialog.DialogExportImg;
import com.mx.gillustrated.R;
import com.mx.gillustrated.util.CommonUtil;
import com.mx.gillustrated.util.DataBakUtil;
import com.mx.gillustrated.util.ServiceUtils;
import com.mx.gillustrated.vo.CardInfo;
import com.mx.gillustrated.vo.GameInfo;

import androidx.appcompat.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Window;
import android.view.inputmethod.BaseInputConnection;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;
import butterknife.BindView;

import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import io.reactivex.functions.Consumer;

public class MainActivity extends BaseActivity {

    public int mGameType = 0; //游戏类别
    private MainActivityHeader mMainActivityHeader;
    private MainActivityTop mMainActivityTop;
    private MainActivityListView mMainActivityListView;
    Handler mainHandler = new MainHandler(this);

    @BindView(R.id.etPinyin) EditText etPinyin;
    @BindView(R.id.btnShowEvents)
    ImageButton btnEvents;

    @OnClick(R.id.btnShowEvents)
    void onBtnShowEventsClick(){
        Intent intentEvent = new Intent(MainActivity.this, EventsActivity.class);
        intentEvent.putExtra("game", mGameType);
        startActivity(intentEvent);
    }

    @BindView(R.id.btnShowMenu) ImageButton btnMenu;

    @OnClick(R.id.btnShowMenu)
    void onBtnShowMenuClick(){
        BaseInputConnection mInputConnection = new BaseInputConnection(btnMenu, true);
        KeyEvent down = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_MENU);
        mInputConnection.sendKeyEvent(down);
        KeyEvent up = new KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_MENU);
        mInputConnection.sendKeyEvent(up);
    }

    @OnTextChanged(R.id.etPinyin)
    void onEtPinyinChanged(){
        boolean breaks = etPinyin.getTag() != null && (boolean) etPinyin.getTag();//防止change事件二次检索
        if(breaks){
            etPinyin.setTag(false);
            return;
        }
        String input = etPinyin.getText().toString();
        mMainActivityListView.searchData(input);
    }

    @OnClick(R.id.btnAdd)
    void btnAddClickListener(){
        Intent intent = new Intent(MainActivity.this, AddCardActivity.class);
        intent.putExtra("game", mGameType);
        startActivity(intent);
    }

   /* test call jni method

    static {
        System.loadLibrary("hello-jni");
    }
    public native String  stringFromJNI();*/

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		requestWindowFeature(Window.FEATURE_NO_TITLE);
		setContentView(R.layout.activity_main);
        mGameType = getIntent().getIntExtra("game", CommonUtil.getGameType(this));
        ButterKnife.bind(this);

        //Log.d("NATIVE ",  stringFromJNI());

        mMainActivityHeader = new MainActivityHeader(this, new MainActivityHeader.HeaderHandle(){
            @Override
            public void onHeaderClick(String column) {
                mMainActivityListView.searchDataOrderby(column);
            }
        }, mGameType);

        mMainActivityTop = new MainActivityTop(this, mOrmHelper, new MainActivityTop.TopHandle(){
            @Override
            public void onSearchData() {
                mMainActivityHeader.setHeaderColor(mMainActivityListView.getOrderBy().split("\\*")[0]);
                CardInfo card = mMainActivityTop.getSpinnerInfo();
                if(card != null) {
                    card.setGameId(mGameType);
                    mMainActivityListView.searchData(card);
                }
            }
            @Override
            public void onRefresh() {
                mMainActivityListView.setOrderBy(null, null);
                etPinyin.setTag(true);
                etPinyin.setText("");
            }

            @Override
            public void onGameTypeChanged(int type) {
                mGameType = type;
                mMainActivityHeader.setResourceController(mGameType);
            }
        });

        mMainActivityListView = new MainActivityListView(this, mOrmHelper, new MainActivityListView.DataViewHandle(){
            @Override
            public void onListItemClick(CardInfo info, int position, int totalCount, int currentPage) {
                Intent intent = new Intent(MainActivity.this,
                        DetailActivity.class);
                intent.putExtra("card", info.getId());
                intent.putExtra("cardSearchCondition", mMainActivityListView.getSearchCondition().getCardSearchParam());
                intent.putExtra("orderBy", mMainActivityListView.getOrderBy());
                intent.putExtra("positon", position);
                intent.putExtra("totalCount", totalCount);
                intent.putExtra("currentPage", currentPage);
                intent.putExtra("spinnerIndexs", mMainActivityTop.getSpinnerSelectedIndexs() );
                startActivity(intent);
            }

            @Override
            public void onSearchCompleted() {
                mProgressDialog.dismiss();
            }

            @Override
            public void onSearchStart() {
                mProgressDialog.show();
            }
        });


		setGameList();

		//temp
//		File fileDirTemp = new File(Environment.getExternalStorageDirectory(),
//				"backup");
//		File fileTemp = new File(fileDirTemp.getPath(), "qingdao.xlsx");
//		 try {
//			CommonUtil.copyBigDataToSD(this, "qingdao.xlsx", fileTemp.getPath());
//		} catch (IOException e) {
//			e.printStackTrace();
//		}

	}



    private void setGameList()
	{
        ServiceUtils.createConnect(new DBCall<List<GameInfo>>() {
            @Override
            public List<GameInfo> enqueue() {
                return mOrmHelper.getGameInfoDao().queryForAll();
            }
        }).doOnNext(new Consumer<List<GameInfo>>() {
            @Override
            public void accept(List<GameInfo> list) throws Exception {
                if(list.size() > 0){
                    mMainActivityTop.setGameList(mGameType, list);
                }else{
                    new Thread() {
                        public void run() {
                            DataBakUtil.getDataFromFiles(mOrmHelper);
                            mainHandler.sendEmptyMessage(1);
                        }
                    }.start();
                }
            }
        }).subscribe();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.menu_main, menu);
		return true;
	}

	@Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        switch(item.getItemId())
        {
        	case  R.id.menu_out :
                DataBakUtil.saveDataToFiles(mOrmHelper);
                Toast.makeText(this, "导出成功", Toast.LENGTH_SHORT).show();
	            break;
        	case  R.id.menu_gamelist :
        		Intent intent = new Intent(MainActivity.this, GameListActivity.class);
    			startActivity(intent);
		        break;
        	case  R.id.menu_header :
        		DialogExportImg.show(this, mMainActivityListView.getDataList().get(0).getId(), mGameType, mainHandler);
		        break;
            case  R.id.menu_eventlist :
                onBtnShowEventsClick();
                break;
            case  R.id.menu_delete :
                onDeleteListData();
                break;
        }
        return true;
    }

    private void onDeleteListData(){
        new AlertDialog.Builder(this)
                .setMessage("确定要删除吗?")
                .setNegativeButton(android.R.string.no, null)
                .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        new Thread(new DeleteRunnable()).start();
                    }
                })
                .show();
    }

    private class DeleteRunnable implements Runnable{

        @Override
        public void run() {
            List<CardInfo> list = mMainActivityListView.getDataList();
            List<Integer> idList = new ArrayList<>();
            for(CardInfo info:list)
                idList.add(info.getId());

            File imagesFileDir = new File(
                    Environment.getExternalStorageDirectory(),
                    MConfig.SD_PATH + "/" + mGameType);
            if (imagesFileDir.exists()) {
                File[] child = imagesFileDir.listFiles();
                for (int i = 0; i < child.length; i++) {
                    String name = child[i].getName();
                    int id = Integer.parseInt(name.split("_")[0].substring(2));
                    if(idList.contains(id))
                        CommonUtil.deleteImage(MainActivity.this, child[i]);
                }
            }

            File imagesHeaderFileDir = new File(
                    Environment.getExternalStorageDirectory(),
                    MConfig.SD_HEADER_PATH + "/" + mGameType);
            if (imagesHeaderFileDir.exists()) {
                File[] child = imagesHeaderFileDir.listFiles();
                for (int i = 0; i < child.length; i++) {
                    String name = child[i].getName();
                    int id = Integer.parseInt(name.split("_")[0]);
                    if(idList.contains(id))
                        CommonUtil.deleteImage(MainActivity.this, child[i]);
                }
            }
            mOrmHelper.getCardInfoDao().delCardsById(idList);
            mainHandler.sendEmptyMessage(5);
        }

    }



    private static class MainHandler extends Handler{

	    private final WeakReference<MainActivity> weakReference;

	    public MainHandler(MainActivity activity){
            weakReference = new WeakReference<>(activity);
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            final MainActivity mainActivity = weakReference.get();
            if (msg.what == 1) {
                mainActivity.setGameList();
            }
            else if(msg.what == 2 || msg.what == 4)
            {
                final int index = msg.what;
                new Thread() {
                    public void run() {
                        CommonUtil.generateHeaderImg(mainActivity, mainActivity.mMainActivityListView.getIdListWithProfile() , mainActivity.mGameType, index != 2);
                        mainActivity.mainHandler.sendEmptyMessage(3);
                    }
                }.start();
            }
            else if (msg.what == 3) {
                Toast.makeText(mainActivity, "生成头像完成", Toast.LENGTH_SHORT).show();
            }
            else if (msg.what == 5) {
                Toast.makeText(mainActivity, "删除完成", Toast.LENGTH_SHORT).show();

            }
        }

    }

}

File: app/src/main/java/com/mx/gillustrated/activity/MainActivity.kt
package com.mx.gillustrated.activity

import android.Manifest
import java.io.File
import java.lang.ref.WeakReference
import java.util.ArrayList
import com.mx.gillustrated.common.DBCall
import com.mx.gillustrated.common.MConfig
import com.mx.gillustrated.component.MainActivityHeader
import com.mx.gillustrated.component.MainActivityListView
import com.mx.gillustrated.component.MainActivityTop
import com.mx.gillustrated.dialog.DialogExportImg
import com.mx.gillustrated.R
import com.mx.gillustrated.util.CommonUtil
import com.mx.gillustrated.util.DataBakUtil
import com.mx.gillustrated.util.ServiceUtils
import com.mx.gillustrated.vo.CardInfo
import com.mx.gillustrated.vo.GameInfo
import androidx.appcompat.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.*
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.Window
import android.view.inputmethod.BaseInputConnection
import android.widget.Toast
import com.mx.gillustrated.databinding.ActivityMainBinding
import com.mx.gillustrated.dialog.FragmentDialogTheme

class MainActivity : BaseActivity() {

    var mGameType = 0 //游戏类别
    private lateinit var mMainActivityHeader: MainActivityHeader
    private lateinit var mMainActivityTop: MainActivityTop
    private lateinit var mMainActivityListView: MainActivityListView
    internal var mainHandler: Handler = MainHandler(this)
    lateinit var binding: ActivityMainBinding

    companion object {
        const val MY_PERMISSIONS_REQUEST = 114
        val permissions:Array<String> = arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_PHONE_STATE)

        private class MainHandler(activity: MainActivity) : Handler(Looper.getMainLooper()) {

            private val weakReference: WeakReference<MainActivity> = WeakReference(activity)

            override fun handleMessage(msg: Message) {
                super.handleMessage(msg)
                val mainActivity:MainActivity = weakReference.get()!!
                when (msg.what) {
                    1 -> {
                        mainActivity.setGameList()
                    }
                    2, 4 -> {
                        val index = msg.what
                        object : Thread() {
                            override fun run() {
                                CommonUtil.generateHeaderImg(mainActivity, mainActivity.mMainActivityListView.idListWithProfile, mainActivity.mGameType, index != 2)
                                mainActivity.mainHandler.sendEmptyMessage(3)
                            }
                        }.start()
                    }
                    3 -> {
                        Toast.makeText(mainActivity, "生成头像完成", Toast.LENGTH_SHORT).show()
                    }
                    5 -> {
                        Toast.makeText(mainActivity, "删除完成", Toast.LENGTH_SHORT).show()
                    }
                    6 -> {
                        mainActivity.mMainActivityTop.setGameList(mainActivity.mGameType,
                                msg.data!!.getParcelableArrayList<GameInfo>("list")!!.toList() )
                    }
                }
            }

        }
    }

    /* test call jni method

    static {
        System.loadLibrary("hello-jni");
    }
    public native String  stringFromJNI();*/


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        mGameType = intent.getIntExtra("game", CommonUtil.getGameType(this))
        requestPermission()
    }

    private fun initListener(){
        binding.btnShowEvents.setOnClickListener {
            val intentEvent = Intent(this@MainActivity, EventsActivity::class.java)
            intentEvent.putExtra("game", mGameType)
            startActivity(intentEvent)
        }
        binding.btnShowMenu.setOnClickListener {
            val mInputConnection = BaseInputConnection(binding.btnShowMenu, true)
            val down = KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_MENU)
            mInputConnection.sendKeyEvent(down)
            val up = KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_MENU)
            mInputConnection.sendKeyEvent(up)
        }
        binding.btnAdd.setOnClickListener {
            val intent = Intent(this@MainActivity, AddCardActivity::class.java)
            intent.putExtra("game", mGameType)
            startActivity(intent)
        }
        binding.etPinyin.addTextChangedListener(object : TextWatcher{
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val breaks = binding.etPinyin.tag != null && binding.etPinyin.tag as Boolean//防止change事件二次检索
                if (breaks) {
                    binding.etPinyin.tag = false
                    return
                }
                val input = binding.etPinyin.text.toString()
                mMainActivityListView.searchData(input)
            }
            override fun afterTextChanged(s: Editable?) {}

        })
    }

    private fun init(){
        //Log.d("NATIVE ",  stringFromJNI());
        initListener()
        mMainActivityHeader = MainActivityHeader(this,
                object : MainActivityHeader.HeaderHandle{
                    override fun onHeaderClick(c: String) {
                        mMainActivityListView.searchDataOrderBy(c)
                    }

                },
                mGameType)

        mMainActivityTop = MainActivityTop(this, binding, mOrmHelper, object : MainActivityTop.TopHandle {
            override fun onSearchData() {
                mMainActivityHeader.setHeaderColor(mMainActivityListView.orderBy.split("\\*".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0])
                val card = mMainActivityTop.spinnerInfo
                if (card != null) {
                    card.gameId = mGameType
                    mMainActivityListView.searchData(card)
                }
            }

            override fun onRefresh() {
                mMainActivityListView.setOrderBy(null, null)
                binding.etPinyin.tag = true
                binding.etPinyin.setText("")
            }

            override fun onGameTypeChanged(type: Int) {
                mGameType = type
                mMainActivityHeader.setResourceController(mGameType)
            }
        })

        mMainActivityListView = MainActivityListView(this, binding, mOrmHelper, object : MainActivityListView.DataViewHandle {
            override fun onListItemClick(info: CardInfo, position: Int, totalCount: Int, currentPage: Int) {
                val intent = Intent(this@MainActivity,
                        DetailActivity::class.java)
                intent.putExtra("card", info.id)
                intent.putExtra("cardSearchCondition", mMainActivityListView.searchCondition.cardSearchParam)
                intent.putExtra("orderBy", mMainActivityListView.orderBy)
                intent.putExtra("positon", position)
                intent.putExtra("totalCount", totalCount)
                intent.putExtra("currentPage", currentPage)
                intent.putExtra("spinnerIndexs", mMainActivityTop.spinnerSelectedIndexes)
                startActivity(intent)
            }

            override fun onSearchCompleted() {
                mProgressDialog.dismiss()
            }

            override fun onSearchStart() {
                mProgressDialog.show()
            }
        })


        setGameList()

        //temp
        //		File fileDirTemp = new File(Environment.getExternalStorageDirectory(),
        //				"backup");
        //		File fileTemp = new File(fileDirTemp.getPath(), "all1.docx");
        //		 try {
        //			CommonUtil.copyBigDataToSD(this, "all1.docx", fileTemp.getPath());
        //		} catch (IOException e) {
        //			e.printStackTrace();
        //		}
    }
    private fun requestPermission() {
        if (!hasPermission()) {
            requestPermissions(permissions, MY_PERMISSIONS_REQUEST)
        } else
            init()
    }

    private fun hasPermission(): Boolean {
        var result = true
        permissions.forEach {
            val res = checkCallingOrSelfPermission(it)
            if (res != PackageManager.PERMISSION_GRANTED)
                result = false
        }
        return result
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            MY_PERMISSIONS_REQUEST -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    init()
                } else {
                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                }
                return
            }
        }
    }

    private fun setGameList() {
        Thread{
            ServiceUtils.createConnect(object : DBCall<List<GameInfo>>() {
                override fun enqueue(): List<GameInfo> {
                    return mOrmHelper.gameInfoDao.queryForAll()
                }
            }).doOnNext { list ->
                if (list.isNotEmpty()) {
                    val msg = Message.obtain()
                    msg.data.putParcelableArrayList("list", list as ArrayList<out Parcelable>)
                    msg.what = 6
                    mainHandler.sendMessage(msg)
                } else {
                    DataBakUtil.getDataFromFiles(mOrmHelper)
                    mainHandler.sendEmptyMessage(1)
                }
            }.subscribe()
        }.start()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        super.onOptionsItemSelected(item)
        when (item.itemId) {
            R.id.menu_out -> {
                DataBakUtil.saveDataToFiles(mOrmHelper)
                Toast.makeText(this, "导出成功", Toast.LENGTH_SHORT).show()
            }
            R.id.menu_gamelist -> {
                val intent = Intent(this@MainActivity, GameListActivity::class.java)
                startActivity(intent)
            }
            R.id.menu_header -> DialogExportImg.show(this, mMainActivityListView.dataList[0].id, mGameType, mainHandler)
            R.id.menu_delete -> onDeleteListData()
            R.id.action_h5-> {
                val intent = Intent(this, WebActivity::class.java)
                startActivity(intent)
            }
            R.id.action_flutter-> {
                val intent = Intent(this, FlutterEventsActivity::class.java)
                intent.action = Intent.ACTION_SEND
                intent.type = "text/plain"
                intent.putExtra("game", mGameType)
                intent.putExtra(Intent.EXTRA_TEXT, "MX")
                startActivity(intent)
            }
            R.id.action_game-> {
                val intent = Intent(this, CultivationActivity::class.java)
                startActivity(intent)
                this.finish()
            }
            R.id.action_theme-> {
                val ft = supportFragmentManager.beginTransaction()
                val prev = supportFragmentManager.findFragmentByTag("dialog_theme")
                if (prev != null) {
                    ft.remove(prev)
                }
                ft.addToBackStack(null)
                // Create and show the dialog.
                val newFragment = FragmentDialogTheme.newInstance()
                newFragment.show(ft, "dialog_theme")
            }
        }
        return true
    }

    private fun onDeleteListData() {
        AlertDialog.Builder(this)
                .setMessage("确定要删除吗?")
                .setNegativeButton(android.R.string.cancel, null)
                .setPositiveButton(android.R.string.ok) { _, _ -> Thread(DeleteRunnable()).start() }
                .show()
    }

    private inner class DeleteRunnable : Runnable {

        override fun run() {
            val list = mMainActivityListView.dataList
            val idList = ArrayList<Int>()
            for (info in list)
                idList.add(info.id)

            val imagesFileDir = File(
                    Environment.getExternalStorageDirectory(),
                    MConfig.SD_PATH + "/" + mGameType)
            if (imagesFileDir.exists()) {
                val child = imagesFileDir.listFiles()
                if (child != null){
                    for (i in child.indices) {
                        val name = child[i].name
                        val id = Integer.parseInt(name.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0].substring(2))
                        if (idList.contains(id))
                            CommonUtil.deleteImage(this@MainActivity, child[i])
                    }
                }
            }

            val imagesHeaderFileDir = File(
                    Environment.getExternalStorageDirectory(),
                    MConfig.SD_HEADER_PATH + "/" + mGameType)
            if (imagesHeaderFileDir.exists()) {
                val child = imagesHeaderFileDir.listFiles()
                if (child != null) {
                    for (i in child.indices) {
                        val name = child[i].name
                        val id = Integer.parseInt(
                            name.split("_".toRegex()).dropLastWhile { it.isEmpty() }
                                .toTypedArray()[0]
                        )
                        if (idList.contains(id))
                            CommonUtil.deleteImage(this@MainActivity, child[i])
                    }
                }
            }
            mOrmHelper.cardInfoDao.delCardsById(idList)
            mainHandler.sendEmptyMessage(5)
        }

    }

}

